File: lib/services/sunglasses_service.dart
import 'package:logging/logging.dart';
import 'package:mongo_dart/mongo_dart.dart';
import 'package:nomac/models/sunglasses/sunglasses_top.dart';

import '../models/sunglasses/sunglasses.dart';
import '../models/user/user.dart';

class SunglassesService {
  SunglassesService(
    Db db,
  ) : _coll = db.collection('sunglasses');

  final _logger = Logger('NOMAC | SunglassesService');
  final DbCollection _coll;

  Future<bool> own(Sunglasses model) async {
    var json = model.toJson();
    json['date'] = model.date;

    try {
      await _coll.insert(json);
      return true;
    } catch (err) {
      _logger.severe(err);
      return false;
    }
  }

  Future<int> ownedCount(String discordId) async {
    return _coll.count(where.eq('victim', discordId));
  }

  Future<int> ownageCount(String discordId) async {
    return _coll.count(where.eq('owned_by', discordId));
  }

  Future<Sunglasses?> mostRecentOwned(String discordId) async {
    final pipeline = AggregationPipelineBuilder()
        .addStage(Match(where.eq('victim', discordId)))
        .addStage(Sort({'date': -1}))
        .addStage(Limit(1))
        .build();

    final res = await _coll.aggregateToStream(pipeline).toList();

    if (res.isEmpty) {
      return null;
    }
    return Sunglasses.fromJson(res.first);
  }

  Future<List<SunglassesTop>> getTop() async {
    final pipeline = AggregationPipelineBuilder()
        .addStage(Group(id: Field('victim'), fields: {'count': Sum(1)}))
        .addStage(Sort({'count': -1}))
        .addStage(Limit(10))
        .build();

    final res = await _coll.aggregateToStream(pipeline).toList();

    return res.map((e) => SunglassesTop.fromJson(e)).toList();
  }
}

File: src/services/sunglasses_service.ts
import { Collection, Db } from "mongodb";
import { autoInjectable, inject } from "tsyringe";
import { Logger } from "pino";
import { ISunglasses } from "../types/sunglasses_type";

@autoInjectable()
export default class SunglassesService {
  private coll: Collection;

  constructor(private db: Db, @inject("pino") private logger: Logger) {
    this.coll = db.collection("sunglasses");
  }

  own = async (victim: string, owned_by: string) => {
    const data: ISunglasses = {
      victim,
      owned_by,
      date: new Date(),
    };

    return await this.coll.insertOne({ data });
  };

  ownedCount = async (discordId: string) => {
    return await this.coll.countDocuments({
      victim: discordId,
    });
  };

  ownageCount = async (discordId: string) => {
    return await this.coll.countDocuments({
      owned_by: discordId,
    });
  };

  mostRecentOwnage = async (discordId: string) => {
    const res = this.coll.aggregate([
      {
        $match: {
          victim: discordId,
        },
      },
      {
        $sort: {
          date: -1,
        },
      },
      {
        $limit: 1,
      },
    ]);

    const arr = (await res.toArray()) as ISunglasses[];

    return arr.length > 0 ? arr[0] : null;
  };

  getTop = async () => {
    const res = this.coll.aggregate([
      {
        $group: {
          _id: "$victim",
          count: {
            $sum: 1,
          },
        },
      },
      {
        $sort: {
          count: -1,
        },
      },
      {
        $limit: 10,
      },
    ]);

    return (await res.toArray()).map((x) => ({
      discordId: x._id,
      count: x.count,
    }));
  };
}
